### DISCUSSION TITLE
is it possible to type hint a decorator that alters a functions argument type?

### DISCUSSION THREAD

**Discussion by AGiantSquid** [microsoft/pyright#5911] [Q&A] [answered]
Consider a function that expects a pydantic object.

```python
class User(BaseModel):
    id: int

def expects_pydantic(x: User) -> int:
    return x.id
```

Now consider a decorator that we can use with functions like this so that they can accept dictionaries.

```python
def accept_dict(fn):
    @wraps(fn)
    def wrapper(x, *args, **kwargs):
        x = User(**x)
        return fn(x, *args, **kwargs)
    return wrapper

@accept_dict
def expects_pydantic(x: User) -> int:
    return x.id

expects_pydantic({'id': 1})
```

Is it possible to correctly typehint the decorator?

I was trying something like this:

```python
StringKeyDict = Dict[str, Any]
P = ParamSpec('P')
R = TypeVar('R')

ModelFunc = Callable[Concatenate[BaseModel, P], R]
DictFunc = Callable[Concatenate[StringKeyDict, P], R]

def accept_dict(fn: ModelFunc) -> DictFunc:
    ...
```

But I get an error when I try to use this:

```python
@accept_dict
def expects_pydantic(x: User):
    print(x.id)
```

```
Type "(x: User) -> int" cannot be assigned to type "ModelFunc[..., Unknown]"
    Parameter 1: type "BaseModel" cannot be assigned to type "User"
      "BaseModel" is incompatible with "User"

```

This error is because the types of the parameters are contravariant. `accept_dict` _could_ do something like the following:

```python
def accept_dict(fn: ModelFunc) -> DictFunc:
    random_model = RandomModel(name='Bob')
    
    fn(random_model)  # <- this will break if fn expects anything other than a generic BaseModel!
    
    return lambda x: x
```

The function logic matches the signature, and would indeed cause a runtime error if run with `expects_pydantic`.

I was hoping by using ParamSpec and Concatenate, I could communicate to Pyright that the function in the signature will not be called internally, but rather passed through, just with a different parameter type. Is this possible?

Entire snippet for reference:

```python
from functools import wraps
from typing import Any, Callable, Dict, TypeVar

from pydantic import BaseModel
from typing_extensions import Concatenate, ParamSpec


StringKeyDict = Dict[str, Any]
P = ParamSpec('P')
R = TypeVar('R')

ModelFunc = Callable[Concatenate[BaseModel, P], R]
DictFunc = Callable[Concatenate[StringKeyDict, P], R]


class User(BaseModel):
    id: int
    

def accept_dict(fn: ModelFunc) -> DictFunc:
    @wraps(fn)
    def wrapper(x, *args: P.args, **kwargs: P.kwargs):
        x = User(**x)
        return fn(x, *args, **kwargs)
    return wrapper


@accept_dict
def expects_pydantic(x: User) -> int:
    return x.id


expects_pydantic({'id': 1})
```

**Comment by erictraut** [comment]
One thing that immediately jumps out at me with your last snippet of code is that you are defining two generic type aliases `ModelFunc` and `DictFunc` that require two type arguments each, but when you later use these type aliases, you're not providing any type arguments.

Another problem is that `ModelFunc` is defined to accept any `BaseModel`, but the implementation of `accept_dict` assumes that it's a `User` object. And another issue is that `x` is not annotated, and you're assigning an incompatible value back to `x`.

Fixing these issues, here's a solution that type checks without error:

```python
from functools import wraps
from typing import Any, Callable, TypeVar
from pydantic import BaseModel
from typing_extensions import Concatenate, ParamSpec


class User(BaseModel):
    id: int


StringKeyDict = dict[str, Any]
P = ParamSpec("P")
R = TypeVar("R")

ModelFunc = Callable[Concatenate[User, P], R]
DictFunc = Callable[Concatenate[StringKeyDict, P], R]


def accept_dict(fn: ModelFunc[P, R]) -> DictFunc[P, R]:
    @wraps(fn)
    def wrapper(x: StringKeyDict, *args: P.args, **kwargs: P.kwargs):
        u = User(**x)
        return fn(u, *args, **kwargs)

    return wrapper


@accept_dict
def expects_pydantic(x: User) -> int:
    return x.id


expects_pydantic({"id": 1})
```

**Comment by AGiantSquid** [comment]
My apologies. I am attempting to distill some slightly more complicated code into something smaller to illustrate my point, and I missed some important details. `accept_dict` should not have had the model type hardcoded as `User`. Here is a revised snippet to hopefully illustrate what I'm looking for:

```python
from functools import wraps
from typing import Any, Callable, Dict, TypeVar, Type

from pydantic import BaseModel
from typing_extensions import Concatenate, ParamSpec


StringKeyDict = Dict[str, Any]
P = ParamSpec('P')
R = TypeVar('R')
BM = TypeVar('BM', bound=BaseModel)

ModelFunc = Callable[Concatenate[BaseModel, P], R]
DictFunc = Callable[Concatenate[StringKeyDict, P], R]


class User(BaseModel):
    id: int


def accept_dict(bm: Type[BM]) -> Callable[[ModelFunc], DictFunc]:
    def outer(fn: ModelFunc) -> DictFunc:
        @wraps(fn)
        def wrapper(x: StringKeyDict, *args: P.args, **kwargs: P.kwargs):
            u = bm(**x)
            return fn(u, *args, **kwargs)
        return wrapper
    return outer


@accept_dict(User)
def expects_pydantic(x: User) -> int:
    return x.id


expects_pydantic({'id': 1})
```

So the `ModelFunc` is supposed to accept a generic `BaseModel`.



"One thing that immediately jumps out at me with your last snippet of code is that you are defining two generic type aliases ModelFunc and DictFunc that require two type arguments each, but when you later use these type aliases, you're not providing any type arguments."

So when I define the alias `ModelFunc = Callable[Concatenate[User, P], R]`, I must supply the type arguments like this? `ModelFunc[P, R]`
I'm going to have to do some more research on that...I'm not quite Groking the syntax.

**Reply by erictraut** [reply]
Yes, generic type aliases require the use of type arguments. If you enable the `reportMissingTypeArgument` check, pyright will tell you when you have omitted type arguments. PEP 484 indicates that when such type arguments are omitted, the type checker should assume they are `Any` (or `...` in the case of a `ParamSpec` or `*tuple[Any, ...]` in the case of a `TypeVarTuple`).

Many Python developers are confused by the syntax for type aliases. [PEP 695](https://peps.python.org/pep-0695/) (which is implemented in Python 3.12) attempts to fix this. 

Here's a version of your more general code that type checks without error. This uses PEP 695 syntax, but you can convert it back to traditional syntax if your code needs to run in versions of Python prior to 3.12.

```python
# pyright: reportMissingTypeArgument=true

from functools import wraps
from typing import Any, Callable
from pydantic import BaseModel
from typing_extensions import Concatenate

StringKeyDict = dict[str, Any]

type ModelFunc[BM: BaseModel, **P, R] = Callable[Concatenate[BM, P], R]
type DictFunc[**P, R] = Callable[Concatenate[StringKeyDict, P], R]

def accept_dict[BM: BaseModel, **P, R](
    bm: type[BM]
) -> Callable[[ModelFunc[BM, P, R]], DictFunc[P, R]]:
    def outer(fn: ModelFunc[BM, P, R]) -> DictFunc[P, R]:
        @wraps(fn)
        def wrapper(x: StringKeyDict, *args: P.args, **kwargs: P.kwargs):
            u = bm(**x)
            return fn(u, *args, **kwargs)

        return wrapper

    return outer

class User(BaseModel):
    id: int

@accept_dict(User)
def expects_pydantic(x: User) -> int:
    return x.id

expects_pydantic({"id": 1})
```

**Reply by AGiantSquid** [reply, marked as answer]
Yeah, I'm not on 3.12 yet, so converting it I think yields this:

```python

from functools import wraps
from typing import Any, Callable, TypeVar

from pydantic import BaseModel
from typing_extensions import Concatenate, ParamSpec

StringKeyDict = dict[str, Any]
P = ParamSpec('P')
R = TypeVar('R')
BM = TypeVar('BM', bound=BaseModel)

ModelFunc = Callable[Concatenate[BM, P], R]
DictFunc= Callable[Concatenate[StringKeyDict, P], R]


def accept_dict(bm: type[BM]) -> Callable[[ModelFunc[BM, P, R]], DictFunc[P, R]]:
    def outer(fn: ModelFunc[BM, P, R]) -> DictFunc[P, R]:
        @wraps(fn)
        def wrapper(x: StringKeyDict, *args: P.args, **kwargs: P.kwargs):
            u = bm(**x)
            return fn(u, *args, **kwargs)

        return wrapper

    return outer


class User(BaseModel):
    id: int


@accept_dict(User)
def expects_pydantic(x: User) -> int:
    return x.id


expects_pydantic({"id": 1})
```

That seems to work for validation! I won't lie, I'm still foggy on why it works with the type aliases, but having this template will be great for my needs. Thank you so much!
